Mathlib/Algebra/Homology/TotalComplex.lean

```diff
@@ -0,0 +1,183 @@
+
+lemma d₁_eq_zero (h : ¬ c₁.Rel i₁ (c₁.next i₁)) :
+    K.d₁ c₁₂ i₁ i₂ i₁₂ = 0 := by
+  dsimp [d₁]
+  rw [K.shape_f _ _ h, zero_comp, smul_zero]
+
+lemma d₂_eq_zero (h : ¬ c₂.Rel i₂ (c₂.next i₂)) :
+    K.d₂ c₁₂ i₁ i₂ i₁₂ = 0 := by
+  dsimp [d₂]
+  rw [HomologicalComplex.shape _ _ _ h, zero_comp, smul_zero]
+
+end
+
+lemma d₁_eq' {i₁ i₁' : I₁} (h : c₁.Rel i₁ i₁') (i₂ : I₂) (i₁₂ : I₁₂) :
+    K.d₁ c₁₂ i₁ i₂ i₁₂ = ComplexShape.ε₁ c₁ c₂ c₁₂ ⟨i₁, i₂⟩ • ((K.d i₁ i₁').f i₂ ≫
+      K.toGradedObject.ιMapObjOrZero (ComplexShape.π c₁ c₂ c₁₂) ⟨i₁', i₂⟩ i₁₂) := by
+  obtain rfl := c₁.next_eq' h
+  rfl
+
+lemma d₁_eq_zero' {i₁ i₁' : I₁} (h : c₁.Rel i₁ i₁') (i₂ : I₂) (i₁₂ : I₁₂)
+    (h' : ComplexShape.π c₁ c₂ c₁₂ ⟨i₁', i₂⟩ ≠ i₁₂) :
+    K.d₁ c₁₂ i₁ i₂ i₁₂ = 0 := by
+  rw [K.d₁_eq' c₁₂ h i₂ i₁₂, K.toGradedObject.ιMapObjOrZero_eq_zero, comp_zero, smul_zero]
+  exact h'
+
+lemma d₁_eq {i₁ i₁' : I₁} (h : c₁.Rel i₁ i₁') (i₂ : I₂) (i₁₂ : I₁₂)
+    (h' : ComplexShape.π c₁ c₂ c₁₂ ⟨i₁', i₂⟩ = i₁₂) :
+    K.d₁ c₁₂ i₁ i₂ i₁₂ = ComplexShape.ε₁ c₁ c₂ c₁₂ ⟨i₁, i₂⟩ • ((K.d i₁ i₁').f i₂ ≫
+      K.toGradedObject.ιMapObj (ComplexShape.π c₁ c₂ c₁₂) ⟨i₁', i₂⟩ i₁₂ h') := by
+  rw [K.d₁_eq' c₁₂ h i₂ i₁₂, K.toGradedObject.ιMapObjOrZero_eq]
+
+lemma d₂_eq' (i₁ : I₁) {i₂ i₂' : I₂} (h : c₂.Rel i₂ i₂') (i₁₂ : I₁₂) :
+    K.d₂ c₁₂ i₁ i₂ i₁₂ = ComplexShape.ε₂ c₁ c₂ c₁₂ ⟨i₁, i₂⟩ • ((K.X i₁).d i₂ i₂' ≫
+    K.toGradedObject.ιMapObjOrZero (ComplexShape.π c₁ c₂ c₁₂) ⟨i₁, i₂'⟩ i₁₂) := by
+  obtain rfl := c₂.next_eq' h
+  rfl
+
+lemma d₂_eq_zero' (i₁ : I₁) {i₂ i₂' : I₂} (h : c₂.Rel i₂ i₂') (i₁₂ : I₁₂)
+    (h' : ComplexShape.π c₁ c₂ c₁₂ ⟨i₁, i₂'⟩ ≠ i₁₂) :
+    K.d₂ c₁₂ i₁ i₂ i₁₂ = 0 := by
+  rw [K.d₂_eq' c₁₂ i₁ h i₁₂, K.toGradedObject.ιMapObjOrZero_eq_zero, comp_zero, smul_zero]
+  exact h'
+
+lemma d₂_eq (i₁ : I₁) {i₂ i₂' : I₂} (h : c₂.Rel i₂ i₂') (i₁₂ : I₁₂)
+    (h' : ComplexShape.π c₁ c₂ c₁₂ ⟨i₁, i₂'⟩ = i₁₂) :
+    K.d₂ c₁₂ i₁ i₂ i₁₂ = ComplexShape.ε₂ c₁ c₂ c₁₂ ⟨i₁, i₂⟩ • ((K.X i₁).d i₂ i₂' ≫
+    K.toGradedObject.ιMapObj (ComplexShape.π c₁ c₂ c₁₂) ⟨i₁, i₂'⟩ i₁₂ h') := by
+  rw [K.d₂_eq' c₁₂ i₁ h i₁₂, K.toGradedObject.ιMapObjOrZero_eq]
+
+/-- The total complex of a bicomplex. -/
+noncomputable def total : HomologicalComplex C c₁₂ where
+  X := K.toGradedObject.mapObj (ComplexShape.π c₁ c₂ c₁₂)
+  d i₁₂ i₁₂' := GradedObject.descMapObj _ (ComplexShape.π c₁ c₂ c₁₂)
+    (fun ⟨i₁, i₂⟩ _ => K.d₁ c₁₂ i₁ i₂ i₁₂' + K.d₂ c₁₂ i₁ i₂ i₁₂')
+  shape i₁₂ i₁₂' h₁₂ := by
+    ext ⟨i₁, i₂⟩ h
+    have w₁ : K.d₁ c₁₂ i₁ i₂ i₁₂' = 0 := by
+      by_cases h₁ : c₁.Rel i₁ (c₁.next i₁)
+      · rw [K.d₁_eq_zero' c₁₂ h₁ i₂ i₁₂']
+        intro h₂
+        exact h₁₂ (by simpa only [← h, ← h₂] using ComplexShape.rel_π₁ c₂ c₁₂ h₁ i₂)
+      · exact d₁_eq_zero _ _ _ _ _ h₁
+    have w₂ : K.d₂ c₁₂ i₁ i₂ i₁₂' = 0 := by
+      by_cases h₁ : c₂.Rel i₂ (c₂.next i₂)
+      · rw [K.d₂_eq_zero' c₁₂ i₁ h₁ i₁₂']
+        intro h₂
+        exact h₁₂ (by simpa only [← h, ← h₂] using ComplexShape.rel_π₂ c₁ c₁₂ i₁ h₁)
```

Comment:
Do you think you can deduplicate these two parts?

Reply:
I will try to experiment with a slightly different design: introducing two horizontal and vertical operators on `total.X` (rather than on individual summands). I do not think there is anything that could be deduplicated here and in the proof of `d_comp_d`, but it may give slightly better looking proofs.